content/blog/withr-3-0-0/index.Rmd

```diff
@@ -0,0 +1,226 @@
+
+- The other missing piece was being able to inspect the contents of the exit hook. The `sys.on.exit()` R helper was created for this purpose but was affected by a bug that prevented it from working inside functions.
+
+We contributed two changes to R 3.5.0 that filled these missing pieces, fixing the `sys.on.exit()` bug and adding an `after` argument to `on.exit()` to allow last-in first-out ordering.
+
+Until now, we haven't been able to leverage these contributions because of our policy of [supporting the current and previous four versions of R](https://www.tidyverse.org/blog/2019/04/r-version-support). Now that enough time has passed, it was time for a rewrite! `withr::defer()`, our version of `on.exit()` that uses better defaults and allows cleaning up resources non-locally (ironically an essential feature for implementing `local_` functions) is now able to be implemented as a simple wrapper around `on.exit()`.
+
+One benefit of the rewrite is that mixing withr tools and `on.exit()` in the same function now correctly interleaves cleanup:
+
+```{r}
+local({
+  on.exit(print(1))
+
+  withr::defer(print(2))
+
+  on.exit(print(3), add = TRUE, after = FALSE)
+
+  withr::defer(print(4))
+
+  print(5)
+})
+```
+
+But the main benefit is increased performance. Here is how `defer()` compared to `on.exit()` in the previous version:
+
+```{r, eval = FALSE}
+base <- function() on.exit(NULL)
+withr <- function() defer(NULL)
+
+# withr 2.5.2
+bench::mark(base(), withr(), check = FALSE)[1:8]
+#> # A tibble: 2 × 8
+#>   expression      min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
+#>   <bch:expr> <bch:tm> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>
+#> 1 base()            0   82ns  6954952.        0B    696.   9999     1
+#> 2 withr()      26.2µs 27.9µs    35172.    88.4KB     52.8  9985    15
+```
+
+withr 3.0.0 has now caught up to `on.exit()` quite a bit:
+
+```{r, eval = FALSE}
+# withr 3.0.0
+bench::mark(base(), withr(), check = FALSE)[1:8]
+#> # A tibble: 2 × 8
+#>   expression      min median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
+#>   <bch:expr> <bch:tm> <bch:>     <dbl> <bch:byt>    <dbl> <int> <dbl>
+#> 1 base()            0   82ns  7329829.        0B       0  10000     0
+#> 2 withr()      2.95µs  3.4µs   280858.        0B     225.  9992     8
+```
+
+Of course `on.exit()` is still much faster, in part because `defer()` supports more features (more on that below), but mostly because `on.exit` is a primitive function whereas `defer()` is implemented as a normal R function. That said we hope that we now have made `defer()` (and the `local_` and `with_` functions that use it) sufficiently fast to be used even in performance-critical micro-tools.
```

Comment:
```suggestion
Of course `on.exit()` is still much faster, in part because `defer()` supports more features (more on that below), but mostly because `on.exit` is a primitive function whereas `defer()` is implemented as a normal R function. That said, we hope that we now have made `defer()` (and the `local_` and `with_` functions that use it) sufficiently fast to be used even in performance-critical micro-tools.
```